### PR TITLE
CL: Swap Non Vault Funds (#782)

### PR DESCRIPTION
**Pull request by magiodev**

### CODE REVIEW

File: smart-contracts/osmosis/contracts/cl-vault/src/vault/swap.rs [hunk 1]
```diff
@@ -85,6 +77,42 @@ pub fn execute_swap_non_vault_funds(
...
+    forced_swap_route: Option<Vec<SwapAmountInRoute>>,
+    twap_window_seconds: Option<u64>,
+) -> Result<CosmosMsg, ContractError> {
+    let vault_config = VAULT_CONFIG.load(deps.storage)?;
+    let dex_router = DEX_ROUTER.may_load(deps.storage)?;
+
+    let twap_price = get_twap_price(
+        &deps.querier,
+        env.block.time,
+        twap_window_seconds.unwrap_or_default(),
```
Comment (lubkoll) [inline smart-contracts/osmosis/contracts/cl-vault/src/vault/swap.rs:95]: Is that correct, i.e. in case we don't provide that from the outside, then we have a twap_window of 0 now. Shouldn't that also be some default-value (like the 24 hardcoded in range.rs)?

File: smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs [hunk 2]
```diff
@@ -82,21 +82,22 @@ fn test_autocompound_with_rewards_swap_non_vault_funds() {
...
+    // BEFORE USERS DEPOSITS
 
     // Keep track of the total refunded amount on token0 from user deposits
     let mut refund0_amount_total = Uint128::zero();
     let mut refund1_amount_total = Uint128::zero();
-
     // Keep track of the total minted shares from deposits
     let mut total_minted_shares_from_deposits = Uint128::zero();
+
+    // This will be useful after all deposits to assert the contract balance
```
Comment (lubkoll) [inline smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs:93]: plz rm comment, that does not help reading the test code

File: smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs [hunk 3]
```diff
@@ -266,98 +268,113 @@ fn test_autocompound_with_rewards_swap_non_vault_funds() {
...
+        initial_rewards_balance.to_string(),
     );
 
     // SWAP NON VAULT ASSETS BEFORE AUTOCOMPOUND ASSETS
 
-    // Swap non vault funds to vault funds
-    // 50000000000ustride to 49500000000uatom as spot price 1.0 less swap_fees
-    // 50000000000ustride to 49500000000uosmo as spot price 1.0 less swap_fees
-    let expected_amount_rewards = 49500000000u128;
+    // Calculate the total rewards swap amount using balanced_fee configs (0.01 is coming from osmosis as flat fee)
```
Comment (lubkoll) [inline smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs:280]: What is balanced_fee configs? And why is it important that 0.01 is a "flat" fee?
to me the code after is clearer than the comment which leads to questions that are not relevant for this test case. I'd prefer to remove that

File: smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs [hunk 4]
```diff
@@ -366,24 +383,19 @@ fn test_autocompound_with_rewards_swap_non_vault_funds() {
...
-        expected_refund_base,
-        balances_after_autocompound_base,
-        &deposit_ratio_approx
-    );
-    // Quote about rewards swapped to token quote
-    let balances_after_autocompound_quote =
+    let after_autocompound_quote_balance =
         get_balance_amount(&app, contract_address.to_string(), DENOM_QUOTE.to_string());
-    assert_eq!(expected_refund_quote, balances_after_autocompound_quote);
+    // Assert that either the base balance is 0, or the quote balance is 0, or both are 0.
```
Comment (lubkoll) [inline smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs:390]: that is easy to understand from the following line. though the comment is confusing as it explicitly mentions the case that both are 0, which is already handled and thus not checked for in the implementation

File: smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/setup.rs [hunk 5]
```diff
@@ -585,7 +587,7 @@ pub fn calculate_deposit_ratio(
     };
 
     // TODO: Compute this based on tokens_provided size
-    let ratio_approx: String = "0.00005".to_string();
+    let ratio_approx: String = "0.02".to_string();
```
Comment (lubkoll) [inline smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/setup.rs:590]: I recall we spoke about this once, but nothing more. What's the reason for this big change?

**Review by lubkoll — CHANGES_REQUESTED** [review]

File: smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs (hunk 3 above)
Comment (magiodev) [reply smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs:274]: Is a flat fee on test_tube https://github.com/osmosis-labs/test-tube/blob/main/packages/osmosis-test-tube/src/module/gamm.rs#L59 as the Gamm module from it is hardcoding 0.01% swap fee and we cannot change that.

File: smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/setup.rs (hunk 5 above)
Comment (magiodev) [reply smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/setup.rs]: I've been able to revert this to the original value after splitting 50% 50% and tweaking a bit the deposit values. I also deprecated it from the setup.rs and inlined it in the autocompounds.rs which is the only place that needs it.

File: smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs [hunk 6]
```diff
@@ -277,16 +270,16 @@ fn test_autocompound_with_rewards_swap_non_vault_funds() {
 
     // SWAP NON VAULT ASSETS BEFORE AUTOCOMPOUND ASSETS
 
-    // Calculate the total rewards swap amount using balanced_fee configs (0.01 is coming from osmosis as flat fee)
+    // Calculate the total rewards swap amount using 0.01 as swap_fee as osmosis_test_tube hardcodes it this way: https://github.com/osmosis-labs/test-tube/blob/main/packages/osmosis-test-tube/src/module/gamm.rs#L59
```
Comment (lubkoll) [inline smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs:273]: fine if you keep it. I still think it is redundant. Of course we have to take into account the fees as osmosis does

File: smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs [hunk 7]
```diff
@@ -322,27 +315,24 @@ fn test_autocompound_with_rewards_swap_non_vault_funds() {
...
     assert_eq!(0u128, after_swap_rewards_balance);
 
     let after_swap_base_balance =
         get_balance_amount(&app, contract_address.to_string(), DENOM_BASE.to_string());
     // Assert vault position tokens balances increased accordingly to the swapped funds from DENOM_REWARD to DENOM_BASE and DENOM_QUOTE
-    assert_approx_eq!(
+    assert_eq!(
         after_deposit_base_balance
-            .checked_add(rewards_swap_amount.into())
+            .checked_add(swap_token_in_amount.into())
```
Comment (lubkoll) [inline smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs:325]: that looks now totally wrong. After a swap I expect `before_swap_balance - swap_token_in_amount = after_swap_balance` and `before_swap_balance + swap_out_amount = after_swap_balance`

File: smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs [hunk 8]
```diff
@@ -134,9 +130,8 @@ fn test_autocompound_with_rewards_swap_non_vault_funds() {
                     .unwrap()
                     .u128(),
                 expected_refund0,
-                &deposit_ratio_approx
+                APPROX_EQ_FACTOR
```
Comment (lubkoll) [inline smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs:133]: fine for now, but it would have been better to do that in a separate PR. Here it is not necessary and added another 8 files to the PR

**Review by lubkoll — CHANGES_REQUESTED** [review]

File: smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs (hunk 7 above)
Comment (magiodev) [reply smart-contracts/osmosis/contracts/cl-vault/tests/test-tube/autocompound.rs]: Yes you right, the math was right as here, working with the assump[tion of spot price = 1.0 we could do that. But the nomenclature was clearly wrong. Fixed!

**Review by lubkoll — DISMISSED** [review]

File: smart-contracts/osmosis/contracts/cl-vault/src/contract.rs [hunk 9]
```diff
@@ -214,8 +216,27 @@ pub fn reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractEr
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
-    MIGRATION_STATUS.save(deps.storage, &MigrationStatus::Open)?;
+pub fn migrate(deps: DepsMut, _env: Env, msg: MigrateMsg) -> Result<Response, ContractError> {
```
Comment (LaurensKubat) [inline smart-contracts/osmosis/contracts/cl-vault/src/contract.rs:219]: Add a unit test for this behaviour in the migration please, this also helps in clarifying for the next rollout what you expect to happen here

**Review by LaurensKubat — DISMISSED** [review]
lgtm

File: smart-contracts/osmosis/contracts/cl-vault/src/contract.rs (hunk 9 above)
Comment (magiodev) [reply smart-contracts/osmosis/contracts/cl-vault/src/contract.rs:219]: Done!

**Review by lubkoll — APPROVED** [review]
